### ISSUE TITLE
wrapper scripts break with latest macOS command line tools

### ISSUE THREAD

**Issue by srlantz** [pmodels/mpich#6713]
The MPICH that is distributed by Homebrew (e.g.) for Mac has the following line in mpicc, mpicxx, and mpifort:
```
final_ldflags="  -Wl,-commons,use_dylibs"
```
But it seems this linker option is no longer supported by `ld` in the latest Command Line Tools update from Sept. 18 2023. This has been [noted elsewhere](https://community.intel.com/t5/Intel-Fortran-Compiler/Mac-Xcode-15-0-unknown-options-commons/td-p/1526357). Thus when I try to compile a simple MPI code I get the following error:
```
$ mpifort hello_mpi.f90 -o hello
ld: unknown options: -commons 
collect2: error: ld returned 1 exit status
```
After much effort, I found a workaround. There is an option to `ld`, namely `-ld_classic`, that causes the Mac `ld` to fall back to its previous behavior. Thus the following works:
```
$ mpifort -Wl,-ld_classic hello_mpi.f90 -o hello 
```
I don't know if the recent change in the behavior of`ld` is a bug, or if it is intentional; I am not an Apple developer. So I am unsure of the correct response to this issue. But it's an issue that will surely effect anyone trying to use MPICH on an arm64 Mac. Apparently other Mac architectures use ld_classic by default, as is stated in the output of `ld -v`:
```
$ ld -v          
@(#)PROGRAM:ld  PROJECT:dyld-1015.7
BUILD 18:48:48 Aug 22 2023
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
will use ld-classic for: armv6 armv7 armv7s arm64_32 i386 armv6m armv7k armv7m armv7em
LTO support using: LLVM version 15.0.0 (static support for 29, runtime is 29)
TAPI support using: Apple TAPI version 15.0.0 (tapi-1500.0.12.3)
Library search paths:
Framework search paths:
```

**Comment by hzhou** [comment]
Please reference https://github.com/pmodels/mpich/issues/6684

**Comment by srlantz** [comment]
Thank you. Since the MPICH source already incorporates a fix for this, it looks like forcing Homebrew to build MPICH from source is the way to go here. Therefore I did a test and got the following results:
```
$ brew install --build-from-source mpich    # (16 minutes later...)
$ mpif90 hello_mpi.f90 -o hello
ld: warning: ignoring duplicate libraries: '-lemutls_w', '-lgcc'
$ mpiexec -n 2 ./hello
 process           0 :Hello, world
 process           1 :Hello, world
```
So it works now, though `ld` gives minor warnings when the wrapper is called.

As noted in #https://github.com/pmodels/mpich/issues/6684, it appears that Homebrew has a little catching up to do with its packaging before the change to `-commons` in Xcode 15 is handled correctly. Perhaps my easy workaround with `-Wl,-ld_classic` may be useful to some in the meantime (at some risk of maybe not being entirely correct for arm64).

**Comment by hzhou** [comment]
@srlantz Thanks for the testing and a potential solution! I think those libraries were added by the `gfortran`  and `gcc`. We'll look into how to avoid duplicates in the wrapper scripts.

**Comment by srlantz** [comment]
@hzhou Yes, you are correct, I should have mentioned that I have `gfortran` present in my environment after installing it with Homebrew. The `gcc` is just Apple clang masquerading as gcc (as usual).

**Comment by hzhou** [comment]
Close as the original issue was resolved.
